src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -3679,7 +3679,10 @@ trait Types
       *  This is not the case if `tp` contains type skolems whose
       *  skolemization level is higher than the level of this variable.
       */
-    def isRelatable(tp: Type) = unrelatable(tp).isEmpty
+    def isRelatable(tp: Type) = !tp.exists {
```

Comment:
This `Type.exists` would create a new object every time it is called. Unless the code may be re-entrant, it may be worth creating an object with a mutable `level` variable, that may be set on every call.

Reply:
Wdym by an object with a mutable variable? I could just extract the collector into a value.

Reply:
This could mean creating, in the main scope (outside of the scope where the `level` var is defined), a special collector object like this:
  
```Scala
object IsRelatableCollector extends TypeCollector[Boolean](false) {
  var barLevel: Int = 0
    def apply(tp: Type): Unit =
      if (!result) tp match {
        case TypeRef(_, ts: TypeSkolem, _) if ts.level > barLevel => result = true
        case _ => tp.foldOver(this)
      }
}
```

So that, at the point where you would call `isRelatable`, you would use 
```Scala
def isRelatable(tp: Type): Boolean = {
  IsRelatableCollector.barLevel = level
  ! IsRelatableCollector.collect(tp).result
}
```

Note, however, that this may only work if the code is re-entrant. If you have doubts, or if you want to go ahead as is, I have given the OK.

Reply:
I don't know if it's re-entrant. There is another option - we could collect the maximum level and then compare. But this could potentially do more work on average, because `exists` can short-circuit if I'm not mistaken?

Reply:
`Type.exists` is based on `Type.find`, which does short-circuit: once it has an answer it stops traversing the Type tree. The problem here is that _every_ call to either `Type.find` or `Type.exists` (and thus every call to `isRelatable`) allocates a new visitor-like TypeCollector object, which I care to avoid. Extracting this object would mean that the special collector for `isRelatable` would only be allocated _once_ in the program run (a global object).

IIUC, re-entrant means in this case that there is a control path that could lead you from one call to `.exists` to another one. I don't think that is the case for this exists, so we could  extract it. However, I would like for @retronym to weigh in.

Reply:
I don't think it's possible to recurse. That's a simple fold over the type. I made the changes you proposed (but I swapped the boolean logic to avoid calling negations).

I also did a similar change to `unrelatable` which collects the list of skolems with higher level.

I'm not happy about the boilerplate and code duplication, but this should be the most performant version.

Reply:
I also changed `CollectTypeCollector`  to use a `ListBuffer` and avoid a list `reverse`.
This is a rabbit hole :laughing: